compiler/rustc_codegen_ssa/src/codegen_attrs.rs

```diff
@@ -608,11 +610,58 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, did: LocalDefId) -> CodegenFnAttrs {
     // strippable by the linker.
     //
     // Additionally weak lang items have predetermined symbol names.
-    if let Some((name, _)) = lang_items::extract(attrs)
+    if let Some((name, span)) = lang_items::extract(attrs)
         && let Some(lang_item) = LangItem::from_name(name)
     {
+        let has_no_mangle = codegen_fn_attrs.flags.contains(CodegenFnAttrFlags::NO_MANGLE);
         if WEAK_LANG_ITEMS.contains(&lang_item) {
             codegen_fn_attrs.flags |= CodegenFnAttrFlags::RUSTC_STD_INTERNAL_SYMBOL;
+            if has_no_mangle {
```

Comment:
This code is only running when you are defining a lang item. I think it should prevent all combinations of `#[rustc_std_internal_symbol]` and `#[no_mangle]`. Can you move it below the `lang_items::extract` if?

Reply:
Just to make sure I understand this correctly, you're suggesting to add another if after the `lang_items::extract`:

```rust
if codegen_fn_attrs.flags.contains(CodegenFnAttrFlags::NO_MANGLE) {
  if is_weak_symbol {
   // issue error
  } else 
  if codegen_fn_attrs.flags.contains(CodegenFnAttrFlags::RUSTC_STD_INTERNAL_SYMBOL) {
    // issue warning
  } 
}
```

Reply:
Looking at it a bit more careful, I think keeping the error in the same location would be fine. The warning should be moved out of the lang item if however.

Also I noticed that you didn't make the warning a future incompat warning: https://rustc-dev-guide.rust-lang.org/diagnostics.html#future-incompatible-lints

Reply:
> Looking at it a bit more careful, I think keeping the error in the same location would be fine. The warning should be moved out of the lang item if however.
> 
Got it.

> Also I noticed that you didn't make the warning a future incompat warning: https://rustc-dev-guide.rust-lang.org/diagnostics.html#future-incompatible-lints

Yep, I wasn't sure if this was the indented way, I'll change it. I have two questions here:
- what issue should I reference?
- what should be the reason? `FutureReleaseErrorDontReportInDeps`?

Reply:
`FutureReleaseErrorReportInDeps` should be fine as reason. We don't need to emit a future incompat error for long given that `#[rustc_std_internal_symbol]` is unstable. As for the issue, that would be a newly created tracking issue for this future incompat warning.

Reply:
> On the other hand, the nightly versions already issue linking errors.

Yeah. I think emitting an error immediately for this case is fine.

Reply:
Do you have an example of a language item that would emit a warning? I am not able to find one that would fit the criteria and I want to have a test for it.

Reply:
I think the only lang items for which `#[no_mangle]` is actively harmful are those that are weak lang items or those that are generic. Weak lang items would be a hard error and generic lang items don't allow `#[no_mangle]` already I think.

Reply:
I modified the code to issue just the error.